pkg/cloudfabric-controller/controllerframework/controller_framework.go

```diff
@@ -403,94 +304,51 @@ func (c *ControllerBase) updateCachedControllerInstances(newControllerInstanceMa
+	newCandidates := reassignControllerKeys(newControllerInstancesLocal)
+	if len(c.sortedControllerInstancesLocal) != len(newCandidates) {
+		return true, newCandidates
 	}
 
-	// active controller instance can unlock instance 1 position ahead of it
-	c.tryUnlockControllerInstance(sortedNewControllerInstancesLocal, newPos)
-
-	return
-}
-
-func (c *ControllerBase) tryUnlockControllerInstance(sortedControllerInstances []controllerInstanceLocal, pos int) {
-	klog.V(4).Infof("entering trying unlock controller instance. pos %v. current controller %s", pos, c.controllerName)
-	if pos >= 1 && sortedControllerInstances[pos-1].isLocked {
-		klog.V(4).Infof("Trying to unlock controller %s on position %d", sortedControllerInstances[pos-1].instanceName, pos-1)
-		err := c.unlockControllerInstanceHandler(sortedControllerInstances[pos-1])
-		klog.V(4).Infof("Done unlocking controller %s on position %d. err %v", sortedControllerInstances[pos-1].instanceName, pos-1, err)
-		if err != nil {
-			// TODO - add retry logic
-			klog.Fatalf("Unable to unlock controller %s instance %s. panic for now.", c.controllerType, sortedControllerInstances[pos-1].instanceName)
+	for i := 0; i < len(newCandidates); i++ {
+		if c.sortedControllerInstancesLocal[i].controllerKey != newCandidates[i].controllerKey ||
+			c.sortedControllerInstancesLocal[i].instanceName != newCandidates[i].instanceName {
+			return true, newCandidates
 		}
 	}
+
+	return false, nil
 }
 
 // Assume both old & new controller instances are sorted by controller key
 // return
-// 		isUpdate: 		controller instances were changed
-//		isSelfUpdate:   is current controller instance updated (include lowerbound, upperbound, islocked)
-//		newLowerBound:  new lowerbound for current controller instance
-//		newUpperBound:  new upperbound for current controller instance
-//		newPos:			new position of current controller instance in sorted controller list (sorted by controllerKey - upperbound)
-func (c *ControllerBase) tryConsolidateControllerInstancesLocal(newControllerInstancesLocal []controllerInstanceLocal) (isUpdated bool, isSelfUpdated bool, newLowerbound int64, newUpperbound int64, newPos int) {
-	isUpdated = false
-	if len(c.sortedControllerInstancesLocal) != len(newControllerInstancesLocal) {
-		isUpdated = true
-	}
+// 		isUpdate: 					controller instances were changed
+//		isSelfUpdate:   			is current controller instance updated (include lowerbound, upperbound, islocked)
+//		newLowerBound:  			new lowerbound for current controller instance
+//		newUpperBound:  			new upperbound for current controller instance
+//		newPos:						new position of current controller instance in sorted controller list (sorted by controllerKey - upperbound)
+//		updatedControllerInstances: new sorted controller instances
+//
+// Note 08-06-2020: Simplify the scenario to always evenly redistribute the entire problem space first
+// Will need to revisit and design for long term solution later after release 0.3 (2020/0730 release)
```

Comment:
It is better to create a tracking issue.

Reply:
> It is better to create a tracking issue.

There is one: https://github.com/futurewei-cloud/arktos/issues/63